many: ensure Current in val set tracking matches PinnedAt (if != 0)

MiguelPires opened the pull request:
Some places were setting ValidationSetTracking's Current to the currently active sequence and others were setting it to the latest, if an older sequence was pinned (see: https://github.com/snapcore/snapd/pull/11707). This was reflected in the output of `snap validate`; even when a previous sequence was pinned, the output always showed the latest one. The first commit of this PR changes this so Current is always equal to PinnedAt, when Pinned != 0 (i.e., there is a pinned assertion). The second commit introduces a Sequence() method that checks the PinnedAt before returning the Current sequence number, to be even more robust.

pedronis commented on daemon/api_validate.go:
```diff
@@ -114,9 +114,6 @@ func listValidationSets(c *Command, r *http.Request, _ *auth.UserState) Response
-		if tr.PinnedAt > 0 {
-			sequence = tr.PinnedAt
-		}
 		sets, err := validationSetForAssert(st, tr.AccountID, tr.Name, sequence)
```
we can't do this because there is problematic code in assertstate that violates this. Indeed is that code that I think we should remove first.  Regarding this my preference might be to add a Sequence() method ValidationSetTracking that does this and have other places using that

pedronis replied:
I would still recommend to make this change as well as turning:

validationSetForAssert(st *state.State, accountID, name string, sequence int) => validationSetsForTracked(*state.State, *assertstate.ValidationSetTracking)

MiguelPires replied:
Sounds good :+1:

pedronis reviewed (commented):
the issue is really that we have incosistent behavior in assertstate when setting ValidationSetTracking.Current, sometimes is set the same as pinnedAt!=0, sometimes is set to the latest known sequence no matter what. This are the places/divergent behavior I found as I understand things:

```
diff --git a/overlord/assertstate/assertstate.go b/overlord/assertstate/assertstate.go
index c742e43b92..d8c83c8152 100644
--- a/overlord/assertstate/assertstate.go
+++ b/overlord/assertstate/assertstate.go
@@ -446,6 +446,7 @@ func RefreshValidationSetAssertions(s *state.State, userID int, opts *RefreshAss
                                        return fmt.Errorf("internal error: cannot find assertion %v when refreshing validation-set assertions", headers)
                                }
                                if vs.Current != as.Sequence() {
+                                       // pinnedAt is 0, OK
                                        vs.Current = as.Sequence()
                                        UpdateValidationSet(s, vs)
                                }
@@ -882,6 +883,7 @@ func TryEnforcedValidationSets(st *state.State, validationSets []string, userID
                }
                vs := a.(*asserts.ValidationSet)
                tr := newTracking[i]
+               // XXX LATEST, can be != pinnedAt
                tr.Current = vs.Sequence()
        }
        for _, tr := range newTracking {
@@ -918,6 +920,7 @@ func ApplyEnforcedValidationSets(st *state.State, valsets map[string]*asserts.Va
                        AccountID: vs.AccountID(),
                        Name:      vs.Name(),
                        Mode:      Enforce,
+                       // SAME as pinnedSeq if not 0
                        Current:   vs.Sequence(),
                        // may be 0 meaning no pinning
                        PinnedAt: pinnedSeq,
@@ -978,6 +981,7 @@ func FetchAndApplyEnforcedValidationSet(st *state.State, accountID, name string,
                Mode:      Enforce,
                // note, sequence may be 0, meaning not pinned.
                PinnedAt: sequence,
+               // XXX LATEST, can be != pinnedAt
                Current:  current,
        }
 
@@ -1002,6 +1006,7 @@ func MonitorValidationSet(st *state.State, accountID, name string, sequence int,
                Mode:      Monitor,
                // note, Sequence may be 0, meaning not pinned.
                PinnedAt:  sequence,
+               // SAME as PinnedAt if not 0
                Current:   as.Sequence(),
                LocalOnly: local,
        }
```

MiguelPires commented:
@pedronis I changed Current to reflect the used validation set in `TryEnforcedValidationSets` and `FetchAndApplyEnforcedValidationSet`. The latter was changed to set Current from the latest sequence number in this PR https://github.com/snapcore/snapd/pull/11707 which was opened to fix a bug. I changed it to set Current from to the validation set used in the enforcement and verified the bug doesn't reoccur. I also changed the spread test checks added in that PR because they were verifying that the output of `snap validate` reflected the latest known sequence number and not the one in effect.

pedronis reviewed (commented):
thanks, this seems to fix and simplify the inconsistency, if needed we might add a Latest field at a later point. See my notes about still avoiding needing to pick PinnedAt vs Current outside of assertstate own code

pedronis commented on daemon/api_validate.go:
```diff
@@ -325,8 +315,8 @@ func forgetValidationSet(st *state.State, accountID, name string, sequence int)
 
-func validationSetForAssert(st *state.State, accountID, name string, sequence int) (*snapasserts.ValidationSets, error) {
-	as, err := validationSetAssertFromDb(st, accountID, name, sequence)
+func validationSetForTracked(st *state.State, tr *assertstate.ValidationSetTracking) (*snapasserts.ValidationSets, error) {
```
I think this should s/Set/Sets/ to match what is returned,  looking at it again also maybe s/Tracked/Tracking/  ?

pedronis reviewed (approved):
thanks, naming nitpick

mardy reviewed (approved):
LGTM, thanks!
